CMSgov/mint-app
# Easi 2152/ready for review BackEnd

# EASI-2152

This Branch comprises the Backend component of the ready for review feature. The goal is to merge this into the feature branch before merging into Main as there are front end tests that should be fixed first.


## Changes and Description

- Ready for Review Functionality
    - remove the task status Complete. Updated existing seeded complete entries to other statuses
    - All Statuses are checked with the calc status [function](https://github.com/CMSgov/mint-app/pull/141/files#diff-7e6f764a5488089470f70b2c8a846f87b1115353360c7921edc32928a52a55abR43-R67)
         - This sets the status to in progress if the section has been modified. If someone sets a status to ready to review, and it wasn't in ready for review, it updates who set it, and the time stamp.
    - Each update resolver for a task list section calls this method through [BaseTaskListSectionPreUpdate](https://github.com/CMSgov/mint-app/pull/141/files#diff-04633ab1386b0771c3dd42f5b0d1cd645a697c4033af82468c2089ab91ff5b6aR8-R27)
    - Business Objects that currently use the apply changes logic and the BaseStruct embedded object now call [BaseStructPreUpdate](https://github.com/CMSgov/mint-app/pull/141/files#diff-faf48408e962d49d0d45989eda93afd5122dd19108422abcbf5f8a7f8f0d089bR8-R20). This is also called by BaseTaskListPreUpdate. It handles updating modified_by, and applying changes.
    - Each Task list section has 2 fields added to it, readyForReviewBy and readyForReviewDts
    - GQL input types now have a separate task status enum type, that is used to limit the options of values the front end can ask to set the task status
    
- Embedded Structs
    -  Embedded structs were implemented to 
         - A. Share fields between related objects
         - B. Create more reusable code when combined with interfaces
    -  [BaseStruct](https://github.com/CMSgov/mint-app/pull/141/files#diff-f85e7894e8625e188476190907fc2dc05daa610ca022636aa40b8fa9758041c8)
          - This is the shared fields that almost all business objects have (except for existing models table)
    - [BaseTaskListSection](https://github.com/CMSgov/mint-app/pull/141/files#diff-7e6f764a5488089470f70b2c8a846f87b1115353360c7921edc32928a52a55ab)
           - The shared fields for task list sections, which also embeds BaseStruct in the definition
          



## How to test this change
Functional steps are listed below. I would also love feedback about naming, and design of the interfaces, embedded structs, and functions.


1. `scripts/dev up:backend:debug && scripts/dev db:seed`
2. Update this seeded plan general characteristics section to ready for review, note that it now shows it is ready for review and the ready for review by and ready for review dts fields have been updated
      a. You can also send this query first without setting status to see that it wasn't set before. Also, feel free to test the other available status. Note the ready for review and ready for review by fields only change when the status wasn't already set to ready for review
      
```gql
mutation updatePlanGeneralCharacteristics{
  
  updatePlanGeneralCharacteristics(id:"41a11243-c0c8-4b6c-8a10-e8b0cbaa03b2", changes:{
        # # Page 1
          isNewModel: true
        #   existingModel:
        #   resemblesExistingModel:
        #   resemblesExistingModelWhich:
        #   resemblesExistingModelHow:
        #   resemblesExistingModelNote:
        #   hasComponentsOrTracks:
        #   hasComponentsOrTracksDiffer:
        #   hasComponentsOrTracksNote:

        #   # Page 2
        #   alternativePaymentModel:
        #   alternativePaymentModelTypes:
        #   alternativePaymentModelNote:
          keyCharacteristics: [PART_C,OTHER]
        #   keyCharacteristicsOther:
          keyCharacteristicsNote: "New Note"
        #   collectPlanBids:
        #   collectPlanBidsNote:
        #   managePartCDEnrollment:
        #   managePartCDEnrollmentNote:
        #   planContactUpdated:
        #   planContactUpdatedNote:

        #   # Page 3
        #   careCoordinationInvolved:
        #   careCoordinationInvolvedDescription:
        #   careCoordinationInvolvedNote:
        #   additionalServicesInvolved:
        #   additionalServicesInvolvedDescription:
        #   additionalServicesInvolvedNote:
        #   communityPartnersInvolved:
        #   communityPartnersInvolvedDescription:
        #   communityPartnersInvolvedNote:

        #   # Page 4
        #   geographiesTargeted:
        #   geographiesTargetedTypes:
        #   geographiesTargetedTypesOther:
        #   geographiesTargetedAppliedTo:
        #   geographiesTargetedAppliedToOther:
        #   geographiesTargetedNote:
        #   participationOptions:
        #   participationOptionsNote:
        #   agreementTypes:
        #   agreementTypesOther:
        #   multiplePatricipationAgreementsNeeded:
        #   multiplePatricipationAgreementsNeededNote:

        #   # Page 5
        #   rulemakingRequired:
        #   rulemakingRequiredDescription:
        #   rulemakingRequiredNote:
        #   authorityAllowances:
        #   authorityAllowancesOther:
        #   authorityAllowancesNote:
        #   waiversRequired:
          waiversRequiredTypes: [FRAUD_ABUSE]
        #   waiversRequiredNote:
        status: READY_FOR_REVIEW

  }){
# Page 1
  isNewModel
  existingModel
  resemblesExistingModel
  resemblesExistingModelWhich
  resemblesExistingModelHow
  resemblesExistingModelNote
  hasComponentsOrTracks
  hasComponentsOrTracksDiffer
  hasComponentsOrTracksNote

  # Page 2
  alternativePaymentModel
  alternativePaymentModelTypes
  alternativePaymentModelNote
  keyCharacteristics
  keyCharacteristicsOther
  keyCharacteristicsNote
  collectPlanBids
  collectPlanBidsNote
  managePartCDEnrollment
  managePartCDEnrollmentNote
  planContactUpdated
  planContactUpdatedNote

  # Page 3
  careCoordinationInvolved
  careCoordinationInvolvedDescription
  careCoordinationInvolvedNote
  additionalServicesInvolved
  additionalServicesInvolvedDescription
  additionalServicesInvolvedNote
  communityPartnersInvolved
  communityPartnersInvolvedDescription
  communityPartnersInvolvedNote

  # Page 4
  geographiesTargeted
  geographiesTargetedTypes
  geographiesTargetedTypesOther
  geographiesTargetedAppliedTo
  geographiesTargetedAppliedToOther
  geographiesTargetedNote
  participationOptions
  participationOptionsNote
  agreementTypes
  agreementTypesOther
  multiplePatricipationAgreementsNeeded
  multiplePatricipationAgreementsNeededNote

  # Page 5
  rulemakingRequired
  rulemakingRequiredDescription
  rulemakingRequiredNote
  authorityAllowances
  authorityAllowancesOther
  authorityAllowancesNote
  waiversRequired
  waiversRequiredTypes
  waiversRequiredNote

  readyForReviewBy
  readyForReviewDts
  status
    
  }
  
  
}
```  


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
